### PR TITLE
Feature: Embedded Sources

### LINKED ISSUE #1550: add UX for embedded sources
- [x]  P0 – when embedded source is fetching, show some feedback to user (overlay) @AdityaHegde 
- [x]  P0 – enable source listing in inspector like the YAML-based sources @hamilton 
- [x]  P0 – enable source item in nav to show source viewer, w/ the route the cached name @hamilton 
- [x]  P0 – replace source icon with lightning bolt @hamilton 
- [x]  P0 – duckdb-compliant paths (like `~/path/to/dataset.csv` or `./../path/to/dataset/csv` should be autocached; currently these run the profiling queries against these datasets with pure duckdb, which is quite slow. @AdityaHegde 
- [x] P1 – add model references in embedded source inspector @hamilton 
- [x]  P0 – provide affordance for refreshing the embedded source  @AdityaHegde 
- [x]  P0 – error state: all possible error states with explicit sources should be returned to the frontend @AdityaHegde 
- [x] P0 – if I change the name of a ~source~ model, we need to invalidate the embedded source catalog items – @AdityaHegde 
- [x] P0 – get the "create model from source" right – @hamilton 
- [x]  P3 – autocomplete the URLs based on existing embedded sources – @Hamilton Ulmer might be able to tackle this

Remaining Work
- [x] P0 – paths that have a search param or hash don't appear to work. Try `from "https://s3.amazonaws.com/tmp-gfx-public-data/banned-books20220819/banned_book_data_combined_lists.csv?a"`, which is a link that accidentally has this search param. This may be a general runtime issue @AdityaHegde 
- [x] P0 – cost estimates don't update after data source is fetched, but do on refresh
- [x]  P0 – the syntax is not duckdb-compliant; we should be utilizing only duckdb’s file path & https. Right now, having the `file://` and the `gs://` prefixes will not run in the duckdb cli. Is this intentional? @AdityaHegde  Note: Have kept the `gs://` syntax for now. We can circle back after the 1st release.
- [x] P0 – local files do not group.
- [x] P0 – get the source listing to be minimum-viable – @hamilton 
- [ ]  P0 - documentation around this feature. @magorlick 
- [ ]  P0 - add / change example to include this. @magorlick 
- [x] P1 – fix autocomplete for this feature or remove it entirely
- [ ] P0 - Generate a hash from the source's content for ID instead of the current method. The current method will not scale when globs are used.

### PR DESCRIPTION
**Pull request by AdityaHegde**
~closes #1550 – see issue for burndown list.~

This PR is very close. Final issues left before merging this, may challenge the release date for `0.19`:
- [x] fix Go tests - Fixed all issues with this PR. https://github.com/rilldata/rill-developer/pull/1603 will fix the remaining failure. EDIT: turns out this issue is on `main`. Filing separate issue.
- [x] P0 – if I edit an embedded source (for instance, changing the end of the URL to select different sources), it does not invalidate the old catalog entries, so I end up seeing these entries 
image
Is an issue on main https://github.com/rilldata/rill-developer/pull/1609
- [x] P0 – if I attempt to run this branch against an existing `stage.db` from `main`, I get an error when I add a YAML source: `Constraint Error: NOT NULL constraint failed: catalog.embeds`

### CODE REVIEW

**Review by begelundmuller — CHANGES_REQUESTED** [review]
Didn't check the frontend logic, and only went into limited depth on the reconcile logic changes. I'm a little worried about embed relationship tracking in the catalog entries, especially as it seems to have added extra complexity to reconcile. The reconcile logic is getting quite hard to understand.

Also note the failing lint and test CI.

File: proto/rill/runtime/v1/api.proto
Comment (AdityaHegde) [reply proto/rill/runtime/v1/api.proto:441]: Hmmm. Should be possible to make this work through DAG.

File: proto/rill/runtime/v1/catalog.proto
Comment (AdityaHegde) [reply proto/rill/runtime/v1/catalog.proto:52]: Yes

File: runtime/connectors/connectors.go
Comment (AdityaHegde) [reply runtime/connectors/connectors.go:174]: There is a TODO in the ticket linked. This code was just for the prototype.

File: runtime/services/catalog/migrator/models/models.go
Comment (AdityaHegde) [reply runtime/services/catalog/migrator/models/models.go:71]: We will need better parsing than regex to match functions. We are not planning to support this for the 1st cut.

Comment (AdityaHegde) [reply runtime/services/catalog/migrator/models/models.go:71]: Ideally the parser would parse and send all settings for an embedded source. This would replace `ExtractTableNames` in the future.

File: runtime/pkg/fileutil/fileutil.go
Comment (begelundmuller) [inline runtime/pkg/fileutil/fileutil.go:133]: If `len(path) > 1`, then it must also check that `path[1] == '/'`. Because `~filename.txt` is a valid regular filename.

File: runtime/pkg/fileutil/fileutil.go
Comment (begelundmuller) [inline runtime/pkg/fileutil/fileutil.go:141]: Typo. Maybe clarify like `"cannot expand '~' in path %q because the current user doesn't have a home directory"`

File: runtime/services/catalog/migrator/models/models.go
Comment (begelundmuller) [inline runtime/services/catalog/migrator/models/models.go:88]: This replacement makes me a little nervous since it can also impact other strings. E.g. in:
```
select *, 'Found in file "s3://bucket/file.csv"' as log from "s3://bucket/file.csv"
```
A quick way to make it safer would be to do the replace on the full original regex match (e.g. `FROM "s3://bucket/file.csv"` including whitespace is replaced to `FROM generated_name`). Could be done by returning a complex result from `ExtractTableNames`.

Not sure if worth it, let me know what you think

File: runtime/services/catalog/migrator/sources/embedded_sources.go
Comment (begelundmuller) [inline runtime/services/catalog/migrator/sources/embedded_sources.go:17]: `ParseEmbeddedSource`?

**Review by begelundmuller — DISMISSED** [review]
Other changes look good :)

File: runtime/pkg/fileutil/fileutil.go
Comment (AdityaHegde) [reply runtime/pkg/fileutil/fileutil.go:133]: Ah got it. Was wondering what this use case is for.

File: runtime/services/catalog/migrator/models/models.go
Comment (AdityaHegde) [reply runtime/services/catalog/migrator/models/models.go:88]: I am not sure how much this case is encountered. One way to do it is to replace the new name in all the matches. The right way to do it using parsers. @magorlick @hamilton any thoughts?

File: web-common/src/features/models/workspace/inspector/ModelInspectorModelProfile.svelte
Comment (djbarnwal) [inline web-common/src/features/models/workspace/inspector/ModelInspectorModelProfile.svelte:163]: Let's move this to inside the script

File: web-common/src/features/models/workspace/inspector/ModelInspectorModelProfile.svelte
Comment (djbarnwal) [inline web-common/src/features/models/workspace/inspector/ModelInspectorModelProfile.svelte:178]: Let's move this to inside the script as well

File: web-common/src/features/models/workspace/inspector/EmbeddedSourceReferences.svelte
Comment (djbarnwal) [inline web-common/src/features/models/workspace/inspector/EmbeddedSourceReferences.svelte:68]: This stale comment can be removed

File: web-common/src/features/models/workspace/inspector/ModelInspectorModelProfile.svelte
Comment (djbarnwal) [inline web-common/src/features/models/workspace/inspector/ModelInspectorModelProfile.svelte:159]: Nit: replace `text-ellipsis overflow-hidden whitespace-nowrap` here and above with `truncate`
https://tailwindcss.com/docs/text-overflow#truncate

File: web-local/src/lib/components/navigation/NavigationEntry.svelte
Comment (djbarnwal) [inline web-local/src/lib/components/navigation/NavigationEntry.svelte:102]: Use `class: truncate={!$$slots["name"]}` as a replacement for 100-102

**Review by djbarnwal — CHANGES_REQUESTED** [review]
Have left some suggestions. Mostly nits and code cleanup.

File: runtime/services/catalog/migrator/models/models.go
Comment (hamilton) [reply runtime/services/catalog/migrator/models/models.go:88]: No strong feelings; I see hte argument in both directions. In the interest of time, I'd like to get this merged for v20 so I think what we have is acceptable for now. If someone encounters a bug by doing this, we'll address then. I doubt we'll need to before we have a SQL parser.

**Comment by hamilton** [comment]
Believe this PR is ready to be merged, but it appears the Go tests aren't passing. Let's wait until @AdityaHegde is back online and merge this Tuesday morning his time.

**Review by hamilton — APPROVED** [review]

**Pull request merged**
